torch/fx/experimental/symbolic_shapes.py

```diff
@@ -138,6 +138,54 @@ def is_concrete_int(a: Union[int, SymInt]):
 
     return False
 
+def canonicalize_bool_expr(expr: sympy.Expr):
+    r""" Canonicalize a boolean expression by transforming it into a lt / le
+    inequality and moving all the non-constant terms to the rhs.
+    We canonicalize And / Ors / Not via cnf and then canonicalize their subexpr
+    recursively
+    nb. sympy.Rel.canonical is not good enough https://github.com/sympy/sympy/issues/25924
+
+    Args:
+        expr (sympy.Expr): Expression to canonicalize
+    """
+    # Canonicalise an inequality by transforming it into a lt / le
+    # inequality and moving all the non-constant terms to the rhs
+    # We canonicalise And / Ors / Not via cnf
+    # nb. Relational.canonical in sympy is broken
+    # https://github.com/sympy/sympy/issues/25924
+
+    if not isinstance(expr, (sympy.Rel, sympy.And, sympy.Or, sympy.Not, sympy.Eq, sympy.Ne)):
+        return expr
+
+    if isinstance(expr, (sympy.And, sympy.Or, sympy.Not)):
+        expr = sympy.logic.boolalg.to_cnf(expr)
```

Comment:
1) Can this logical expression simplify to False/True?

Reply:
1. I wouldn't expect so. 2., I think it's fine with `simplify=False` as `simplify=True` looks like it does things that are too fancy and potentially expensive.

Chances are the original expression is already very close to being in CNF form, so this would just push `Neg`s inside and change a few ors for ands and and a few ands for ors.

Reply:
I triggered a benchmark run and everything looks reasonably within noise. If this ever becomes an issue, we can implement our own simplified `to_cnf`.